### PR TITLE
ICU-20602 add LocaleBuilder::copyErrorTo

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20602
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/localebuilder.cpp [hunk 1]
```diff
@@ -433,4 +433,9 @@ Locale LocaleBuilder::build(UErrorCode& errorCode)
     return product;
 }
 
+UBool LocaleBuilder::copyErrorTo(UErrorCode &outErrorCode) {
```
Comment (sffc) [inline icu4c/source/common/localebuilder.cpp]: This does not

> Preserves older error codes in the outErrorCode.

Please add

```cpp

        if (U_FAILURE(outErrorCode)) {
            // Do not overwrite the older error code
            return TRUE;
        }
```

File: icu4c/source/common/localebuilder.cpp [hunk 2]
```diff
@@ -433,4 +433,9 @@ Locale LocaleBuilder::build(UErrorCode& errorCode)
...
+    outErrorCode = status_;
+    return U_FAILURE(outErrorCode) ? TRUE : FALSE;
```
Comment (sffc) [inline icu4c/source/common/localebuilder.cpp]: U_FAILURE already returns a UBool.  No need to pass it through another ternary.

File: icu4c/source/test/intltest/localebuildertest.cpp [hunk 3]
```diff
@@ -190,6 +197,9 @@ void LocaleBuilderTest::TestLocaleBuilder() {
         status = U_ZERO_ERROR;
         bld.clear();
         while (true) {
+            status = U_ZERO_ERROR;
+            UErrorCode copyStatus = U_ZERO_ERROR;
+            bld.copyErrorTo(copyStatus);
```
Comment (sffc) [inline icu4c/source/test/intltest/localebuildertest.cpp]: You copy the error before you start mutating the LocaleBuilder.  Wouldn't you want to put this after you call all the various builder methods?

I would also like to see a concrete test where the error code is not just U_ZERO_ERROR.  Give it bad input on purpose, and check that the error code is as expected.  Maybe you already have that, but it's not obvious from looking at the test.

File: icu4c/source/test/intltest/localebuildertest.cpp [hunk 4]
```diff
@@ -55,10 +55,17 @@ void LocaleBuilderTest::runIndexedTest( int32_t index, UBool exec, const char* &
 
 void LocaleBuilderTest::Verify(LocaleBuilder& bld, const char* expected, const char* msg) {
     UErrorCode status = U_ZERO_ERROR;
```
Comment (sffc) [inline icu4c/source/test/intltest/localebuildertest.cpp]: FYI for the future: in tests, you can and should use IcuTestErrorCode from testlog.h.

File: icu4c/source/common/localebuilder.cpp (hunk 2 above)
Comment (FrankYFTang) [reply icu4c/source/common/localebuilder.cpp]: DONE

File: icu4c/source/common/localebuilder.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/common/localebuilder.cpp]: DONE

**Comment by FrankYFTang** [comment]
PTAL

**Comment by FrankYFTang** [comment]
> I don't see:
> 
> > I would also like to see a concrete test where the error code is not just U_ZERO_ERROR. Give it bad input on purpose, and check that the error code is as expected. Maybe you already have that, but it's not obvious from looking at the test.

Inside the LocaleBuilderTest::Verify function

**Comment by sffc** [comment]
Ok.  Trying to read your verify function.  So you have data with "X", and then you have

```cpp
            else if (strcmp("X", method) == 0) {
                if (U_SUCCESS(status)) {
                    errln("FAIL: No error return - test case: %s", actions.c_str());
                }
```

And then you also have

```cpp
            if (status != copyStatus) {
                errln("copyErrorTo not matching");
            }
```

However, between those two lines, you have

```cpp
            if (U_FAILURE(status)) {
                if (strcmp("X", testCase[i]) == 0) {
                    // This failure is expected
                    break;
```

Can you verify that `copyStatus == status` also in the failure case?

**Comment by FrankYFTang** [comment]
> Ok. Trying to read your verify function. So you have data with "X", and then you have
> 
> ```c++
>             else if (strcmp("X", method) == 0) {
>                 if (U_SUCCESS(status)) {
>                     errln("FAIL: No error return - test case: %s", actions.c_str());
>                 }
> ```
> 
> And then you also have
> 
> ```c++
>             if (status != copyStatus) {
>                 errln("copyErrorTo not matching");
>             }
> ```
> 
> However, between those two lines, you have
> 
> ```c++
>             if (U_FAILURE(status)) {
>                 if (strcmp("X", testCase[i]) == 0) {
>                     // This failure is expected
>                     break;
> ```
> 
> Can you verify that `copyStatus == status` also in the failure case?

DONE. PTAL

**Comment by FrankYFTang** [comment]
> It looks like you used the new squash commit button. There was a bug in that button which I fixed yesterday at 9pm PDT, but it looks like you clicked the button before I fixed it.
> 
> Please force-push from your local branch (reset this PR to what it was before you squashed), then you can try the squash flow again.

PTAL

File: icu4c/source/common/unicode/localebuilder.h [hunk 5]
```diff
@@ -276,6 +276,17 @@ class U_COMMON_API LocaleBuilder : public UObject {
      */
     Locale build(UErrorCode& status);
 
+    /**
+     * Sets the UErrorCode if an error occurred while recording sets.
+     * Preserves older error codes in the outErrorCode.
+     * @param outErrorCode Set to an error code if it does not contain one already
+     *                  and an error occurred while recording sets.
```
Comment (sffc) [inline icu4c/source/common/unicode/localebuilder.h]: Suggested documentation of `@param outErrorCode`:

```
Set to an error code that occurred while setting subtags.
Unchanged if there is no such error or if outErrorCode
already contained an error.
```

Comment (FrankYFTang) [reply icu4c/source/common/unicode/localebuilder.h]: DONE PTAL

**Pull request merged**